File: 1644-lowest-common-ancestor-of-a-binary-tree-ii/1644-lowest-common-ancestor-of-a-binary-tree-ii.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */
class Solution {
    
    int nodesFound = 0;
    
    public TreeNode lowestCommonAncestor(TreeNode root, TreeNode p, TreeNode q) {
        TreeNode lca = findLCA(root, p, q);
        if(nodesFound == 2) return lca;
        return null;
    }
    
    private TreeNode findLCA(TreeNode root, TreeNode p, TreeNode q){
        if(root == null) return null;
        TreeNode left = findLCA(root.left, p, q);
        TreeNode right = findLCA(root.right, p, q);
        if((root.val == p.val) || (root.val == q.val)){
            nodesFound += 1;
            return root;
        }
        return left == null ? right : right == null ? left : root;
    }
}

File: 1644-lowest-common-ancestor-of-a-binary-tree-ii/1644-lowest-common-ancestor-of-a-binary-tree-ii.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val) {
 *     this.val = val;
 *     this.left = this.right = null;
 * }
 */
/**
 * @param {TreeNode} root
 * @param {TreeNode} p
 * @param {TreeNode} q
 * @return {TreeNode}
 */
var lowestCommonAncestor = function(root, p, q) {
    if(!p || !q || !root) return null;
    const lcaFound = { count : 0};
    
    const lca = findLCA(root, p.val, q.val, lcaFound);
    return lcaFound.count === 2 ? lca : null;
};

function findLCA(node, p, q, lcaFound){
    if(node === null) return null;
    const left = findLCA(node.left, p, q, lcaFound);
    const right = findLCA(node.right, p, q, lcaFound);
    if(node.val === p || node.val === q){
        lcaFound.count += 1;
        return node;
    }
    if(left && right) return node;
    return left === null ? right : left;
}